### PR TITLE
Java persistence JPA

### PR DESCRIPTION
**Pull request by Mariana-VV**

### CODE REVIEW

File: src/main/java/org/bookstore/model/Book.java [hunk 1]
```diff
@@ -11,12 +11,16 @@
...
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
 
 @Getter
 @Setter
 @Entity
 @NoArgsConstructor
 @AllArgsConstructor
+@SQLDelete(sql = "update books set is_deleted = true where id=?")
+@Where(clause = "is_deleted=false")
```
Comment (okuzan) [inline src/main/java/org/bookstore/model/Book.java]: Where is deprecated, use SQLrestriction

File: src/main/java/org/bookstore/model/Book.java [hunk 2]
```diff
@@ -32,7 +36,10 @@ public class Book {
     private BigDecimal price;
     private String description;
     private String coverImage;
+    @Column(nullable = false)
+    private boolean isDeleted = false;
 
+    //    @ToString.Exclude
```
Comment (okuzan) [inline src/main/java/org/bookstore/model/Book.java]: don't leave commented code, just remove it, we can get it from git when needed

File: src/main/java/org/bookstore/repository/BookRepository.java [hunk 3]
```diff
@@ -1,15 +1,14 @@
...
+import org.springframework.data.jpa.repository.JpaRepository;
 
-public interface BookRepository {
-    Book save(Book book);
+public interface BookRepository extends JpaRepository<Book, Long> {
 
-    List<Book> findAll();
+    // @Query("from Book b where upper(b.title) like upper(:title) ")
+    // @Query(value = "select * from books where upper(title) like upper(:title) ",
+    // nativeQuery = true)
```
Comment (okuzan) [inline src/main/java/org/bookstore/repository/BookRepository.java]: same

File: src/main/resources/db/changelog/changes/01-create-books-table.yaml [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+                  type: varchar(255)
+              - column:
+                  name: cover_image
+                  type: varchar(255)
+              - column:
+                  name: is_deleted
+                  type: boolean
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
```
Comment (okuzan) [inline src/main/resources/db/changelog/changes/01-create-books-table.yaml]: this file doesn't end with an empty line

File: pom.xml [hunk 5]
```diff
@@ -34,6 +34,18 @@
...
+            <artifactId>liquibase-core</artifactId>
+            <version>${liquibase.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.liquibase</groupId>
+            <artifactId>liquibase-maven-plugin</artifactId>
+            <version>${liquibase.version}</version>
+        </dependency>
+
+
```
Comment (okuzan) [inline pom.xml]:
```suggestion
```

File: src/main/java/org/bookstore/controller/BookController.java [hunk 6]
```diff
@@ -29,13 +33,24 @@ public List<BookDto> findAll() {
...
+    @ResponseStatus(HttpStatus.NO_CONTENT)
     @GetMapping("/{id}")
     public BookDto getBookById(@PathVariable Long id) {
         return bookService.getBookById(id);
     }
 
-    @GetMapping("/by-title")//by-name?title=bob
-    public List<BookDto> getAllByName(@RequestParam String title) {
-        return bookService.getAllByName(title);
+    @GetMapping("/by-title")//http://localhost:8080/api/books/by-title?title=matrix
```
Comment (okuzan) [inline src/main/java/org/bookstore/controller/BookController.java]:
```suggestion
    @GetMapping
```
In RESTful design, the URL should ideally represent a resource and not include verbs or actions. RESTful URLs should be nouns and represent the resources being accessed. In this case, the URL is designed to find books by title, but it includes a query parameter that specifies the title.

A more RESTful approach would be to use a URL that represents the books resource and include the title as part of the resource path or as a query parameter in a more resource-centric way.

**Review by okuzan — CHANGES_REQUESTED** [review]

File: src/main/java/org/bookstore/controller/BookController.java [hunk 7]
```diff
@@ -29,13 +33,24 @@ public List<BookDto> findAll() {
...
+    @ResponseStatus(HttpStatus.NO_CONTENT)
     @GetMapping("/{id}")
     public BookDto getBookById(@PathVariable Long id) {
         return bookService.getBookById(id);
     }
 
-    @GetMapping("/by-title")//by-name?title=bob
-    public List<BookDto> getAllByName(@RequestParam String title) {
-        return bookService.getAllByName(title);
+    @GetMapping("/title")
```
Comment (MAMentorFX) [inline src/main/java/org/bookstore/controller/BookController.java]: Not the best endpoint. localhost:8080/title?title="Sherlock". there are two `title` in uri

File: src/main/java/org/bookstore/service/impl/BookServiceImpl.java [hunk 8]
```diff
@@ -35,8 +35,23 @@ public BookDto getBookById(Long id) {
         return bookMapper.toDto(book);
     }
 
+    public List<BookDto> findAllByTitle(String title) {
+        return bookRepository.findAllByTitleIgnoreCase(title).stream()
+                .map(bookMapper::toDto)
+                .toList();
```
Comment (MAMentorFX) [inline src/main/java/org/bookstore/service/impl/BookServiceImpl.java]: Mapstruct supports mapping collections from the box

File: src/main/java/org/bookstore/service/impl/BookServiceImpl.java [hunk 9]
```diff
@@ -35,8 +35,23 @@ public BookDto getBookById(Long id) {
...
+
+    @Override
+    public void deleteById(Long id) {
+        bookRepository.deleteById(id);
+    }
+
     @Override
-    public List<BookDto> getAllByName(String title) {
-        return bookMapper.toDtoList(bookRepository.getAllByName(title));
+    public BookDto update(Long id, CreateBookRequestDto requestDto) {
```
Comment (MAMentorFX) [inline src/main/java/org/bookstore/service/impl/BookServiceImpl.java]: @Transactional. You make more than one call to DB

**Review by MAMentorFX — CHANGES_REQUESTED** [review]

**Review by MentorMateAcademy — APPROVED** [review]

**Pull request merged**